refactor(core): Ensure that `previousOrParentTNode` always belongs to current `TView`.

mhevery opened the pull request:
`previousOrParentTNode` stores current `TNode`. Due to inconsistent
implementation the value stored would sometimes belong to the current
`TView` and sometimes to the parent. We have extra logic which accounts
for it. A better solution is to just ensure that `previousOrParentTNode`
always belongs to current `TNode`. This simplifies the mental model
and cleans up some code.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

mhevery committed:
refactor(core): Ensure that `previousOrParentTNode` always belongs to current `TView`.

`previousOrParentTNode` stores current `TNode`. Due to inconsistent
implementation the value stored would sometimes belong to the current
`TView` and sometimes to the parent. We have extra logic which accounts
for it. A better solution is to just ensure that `previousOrParentTNode`
always belongs to current `TNode`. This simplifies the mental model
and cleans up some code.

alfaproject commented on packages/core/src/render3/state.ts:
```diff
@@ -54,6 +54,7 @@ interface LFrame {
    *
    * This is used in conjunction with `isParent`.
    */
+  // FIXME(misko): should be `TNode|null` (add comment explaining it.)
```
Maybe also rename to `currentTNode`?

mhevery replied:
Done.

mhevery committed:
refactor(core): Remove host `TNode` from `getOrCreateTNode`

Host `TNode` was passed into `getOrCreateTNode` just so that we can
compute weather or not we are a root node. This was needed because
`previousOrParentTNode` could have `TNode` from `TView` other then
current `TView`. This is confusing mental model. Previous change
ensured that `previousOrParentTNode` must always be part of `TView`,
which enabled this change to remove the unneeded argument.

AndrewKushnir commented on packages/core/src/render3/state.ts:
```diff
@@ -404,7 +406,7 @@ export function enterDI(newView: LView, tNode: TNode) {
  * @param tNode Element to which the View is a child of
```
This argument description can be removed as well.

petebacondarwin replied:
Still needs removing.

AndrewKushnir commented on packages/core/src/render3/util/view_utils.ts:
```diff
@@ -207,3 +207,16 @@ export function updateTransplantedViewCount(lContainer: LContainer, amount: 1|-
+ * the root but component views do not. A lot of logic does not expect to see `TNodeType.View` and
+ * crashes on it, so we unwrap it.
+ */
+export function getNonViewFirstChild(tView: TView): TNode|null {
```
I was trying to think of a name for this function. May be something like this: `unwrapTViewFirstChild`?

petebacondarwin replied:
How about `getRootNode(tView: TView): TNode|null`?

This is what it is really doing, right?
The fact that this root node is stored in the `firstChild` property or potentially an unwrapped version of it is just an implementation detail.

mhevery replied:
Function has been deleted.

AndrewKushnir replied:
This function was removed :)

AndrewKushnir reviewed (commented):
Great refactoring and simplification, thanks Misko! 👍

mhevery committed:
refactor(core): Add injector debug information to `LViewDebug`

Extended the `LViewDebug` to display node-injector information for each
node.

petebacondarwin commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -3,7 +3,7 @@
       "uncompressed": {
         "runtime-es2015": 1485,
-        "main-es2015": 141151,
+        "main-es2015": 141711,
```
Can you explain why this commit has added over 500 bytes to the main for hello world?
It seems that the code inside core is actually reduced by this commit. So I assume that the increase in main is due to additional ivy instructions being added now, which were not there before?

mhevery replied:
It is now smaller.

petebacondarwin replied:
141711 - 141151 = 560 bytes

This doesn't seem smaller...

petebacondarwin commented on packages/core/src/render3/i18n/i18n_apply.ts:
```diff
@@ -412,7 +412,7 @@ export function i18nEndFirstPass(tView: TView, lView: LView) {
   // Remove deleted nodes
   let index = rootIndex + 1;
-  while (index <= lastCreatedNode.index - HEADER_OFFSET) {
+  while (lastCreatedNode && index <= lastCreatedNode.index - HEADER_OFFSET) {
```
Although there is a precedent for this approach on line 421 below.

mhevery replied:
fixed

petebacondarwin commented on packages/core/src/render3/instructions/template.ts:
```diff
@@ -37,9 +37,11 @@ function templateFirstCreatePass(
+      createTNode(embeddedTView, null, TNodeType.View, -1, null, null) as TViewNode;
   embeddedTViewNode.injectorIndex = tNode.injectorIndex;
-  embeddedTView.node = embeddedTViewNode;
+  // FIXME(misko): remove `embeddedTView.node'
```
Is it true that `view.node` could be replaced with `getNonViewFirstChild(view)`?

petebacondarwin commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -213,43 +213,44 @@ export function createLView<T>(
  */
 export function getOrCreateTNode(
-    tView: TView, tHostNode: TNode|null, index: number, type: TNodeType.Element, name: string|null,
+    tView: TView, index: number, type: TNodeType.Element, name: string|null,
```
The `@param tHostNode` needs to be removed from the JSDoc.

mhevery replied:
fixed

petebacondarwin commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -213,43 +213,44 @@ export function createLView<T>(
   // Parents cannot cross component boundaries because components will be used in multiple places,
   // so it's only set if the view is the same.
-  const parentInSameView = parent && parent !== tHostNode;
+  // FIXME(misko): This check for `TNodeType.View` should not be needed. But removing it breaks DI,
+  // so more investigation is needed.
```
I think that this might be because we need to reset `previousOrParentTNode` when starting a new `executeTemplate()`.

One case where this fails is when there is a template inside a container and the element where the directive is being injected is the first item in the template.

petebacondarwin replied:
But I admit I don't really understand the meaning/usage of `previousOrParentTNode` as a state value.

petebacondarwin replied:
So my take on this is that `previousOrParentTNode` is whatever the previous loop of the "rendering engine" processed. This is normally the node that is either the previous or parent node of the node being processed now. But when we start rendering a new view this can be the node that was being processed just before we started rendering this view, i.e. the node where this view is declared - I think.

So when we are creating a new view, we really don't want to attach its TNode to a node in another view but this is what happens because `previousOrParentTNode` is just exactly that.

I think that the reason the DI fails if the root node of a view is attached to a node in another via the `parent` property is that the DI will follow up the Node tree looking for injectors rather than bailing out and then looking locally for an injector that is related to this view. Does that sound right?

If this is accurate, then I would say that this check is actually correct, and needed.

mhevery replied:
This code has been deleted.

petebacondarwin reviewed (commented):
In general this looks like an improvement to the current situation - I think. That being said I find the instruction state rather confusing since it is difficult to track when it gets updated.

It does seem concerning though, that a non-urgent refactoring brings in two new FIXME comments. There are quite of few of these in the code base now, and I suspect that they are unlikely to attract much attention after this PR lands.
If there is a concern then I feel it would be better to address it earlier rather than later. (I.e. as part of this PR).

mhevery committed:
refactor(core): Remove reliance on `TNodeType.View`.

`TNodeType.View` was created to support inline views. That feature did
not materialize and we have since removed the instructions for it, leave
 an unneeded `TNodeType.View` which was still used in a very
 inconsistent way. This change no longer created `TNodeType.View` (and
 there will be a follow up chang to completely remove it.)

Also simplified the mental model so that `LView[HOST]`/`LView[T_HOST]`
always point to the insertion location of the `LView`.

mhevery committed:
refactor(core): Remove `TViewNode` as it is no longer used.

Previous commit change the logic to not rely on the `TViewNode` this
change removes it entirely.

mhevery committed:
refactor(core): Rename `TView.node` to `TView.declTNode`.

The value stored in `TView.node` is really the declaration `TNode`,
therefore renaming to make it more explicit.

mhevery committed:
refactor(core): change `getPreviousOrParentTNode` to return `TNode|null`

This change makes `getPreviousOrParentTNode` return `TNode|null` (rather
than just `TNode`) which is more reflective of the reality. The
`getPreviousOrParentTNode` can be `null` upon entering the `LView`.

mhevery committed:
refactor(core): renamed `previousOrParent` to `currentTNode`

The previous name of `previousOrParent` was confusing. Changed the
terminology to `currentTNode`.

mhevery commented on packages/core/src/render3/di.ts:
```diff
@@ -141,41 +141,42 @@ export function bloomAdd(
   }
 
-  hostView[injectorIndex + PARENT_INJECTOR] = parentLoc;
+
```
fixed

mhevery commented:
> It does seem concerning though, that a non-urgent refactoring brings in two new FIXME comments. There are quite of few of these in the code base now, and I suspect that they are unlikely to attract much attention after this PR lands.
> If there is a concern then I feel it would be better to address it earlier rather than later. (I.e. as part of this PR).

Most of these were removed in this PR.

AndrewKushnir commented:
[Initial presubmit](http://test/OCL:331705695:BASE:331705731:1600150671672:f639a2f).

petebacondarwin commented:
In the first commit is there a typo?

> A better solution is to just ensure that `previousOrParentTNode`
always belongs to current ~~`TNode`~~ `TView`.

petebacondarwin commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -9,7 +9,7 @@
 import {ViewEncapsulation} from '../metadata/view';
 import {addToArray, removeFromArray} from '../util/array_utils';
-import {assertDefined, assertDomNode, assertEqual, assertString} from '../util/assert';
+import {assertDefined, assertDomNode, assertEqual, assertSame, assertString} from '../util/assert';
```
This import should not be added in this commit (it is not used). And it is actually removed in a later commit.

petebacondarwin commented on packages/core/test/render3/i18n_spec.ts:
```diff
@@ -473,7 +475,9 @@ describe('Runtime i18n', () => {
       const opCodes = getOpCodes(attrs, () => {
+        ɵɵelementStart(0, 'div');
         ɵɵi18nAttributes(index, 0);
+        ɵɵelementEnd();
```
Can you explain why/how this test now results in the `div` element attributes, whereas it wasn't before? Is it because the structure of the TView has changed?

mhevery replied:
This test was cheating. It was creating a attributes but did not have an element to create them on. I added assert statements and now the tests fails. The fix is to add `div` so that the attributes have someplace to go.

petebacondarwin commented:
Typo in the 2nd commit:

> we can compute ~~weather~~whether or not we are a root node

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -202,7 +202,7 @@ export function getInjectorIndex(tNode: TNode, hostView: LView): number {
  * parent injector initially.
  *
- * Returns a combination of number of `ViewData` we have to go up and index in that `Viewdata`
+ * Returns a combination of number of `LView` we have to go up and index in that `LView`
```
```suggestion
 * Returns the number of `LView` we have to go up and index in that `LView`
```

petebacondarwin replied:
Don't fancy this?

petebacondarwin replied:
Oh I totally didn't get this comment. Looking at it again, and reading the code in more detail, I think it means:

```
Returns a number that is the combination of the number of `LView`s that we have to go up to find the `LView` containing the parent inject AND the index of the injector within that `LView`.
```

petebacondarwin commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -525,6 +525,7 @@ function _queryAllR3(
 function _queryNodeChildrenR3(
     tNode: TNode, lView: LView, predicate: Predicate<DebugElement>|Predicate<DebugNode>,
     matches: DebugElement[]|DebugNode[], elementsOnly: boolean, rootNativeNode: any) {
+  ngDevMode && assertDefined(tNode, 'TNode required');
```
If this could potentially be `undefined` I would expect the function signature to reflect that...

petebacondarwin replied:
Perhaps really, there is some casting (or element access) going on in the code that calls this that means we are not sure...

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -212,16 +212,17 @@ export function getParentInjectorLocation(tNode: TNode, view: LView): RelativeIn
   // or container), so this loop will be skipped, but we must keep the loop here to support
   // the rarer case of deeply nested <ng-template> tags or inline views.
-  let hostTNode = view[T_HOST];
+  let declTNode = view[TVIEW].node;
```
🎉

petebacondarwin commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -10,7 +10,7 @@ import {AttributeMarker, ComponentTemplate} from '..';
 import {Sanitizer} from '../../sanitization/sanitizer';
 import {KeyValueArray} from '../../util/array_utils';
-import {assertDefined} from '../../util/assert';
+import {assertDefined, assertNotEqual} from '../../util/assert';
```
This import is not used in this commit so should not be added here. In fact it is removed in a subsequent commit.

petebacondarwin commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -250,7 +250,7 @@ function createTNodeAtIndex(
   const parentInSameView = parent !== null && parent.type !== TNodeType.View;
   const tParentNode = parentInSameView ? parent as TElementNode | TContainerNode : null;
```
This `tParentNode` is no longer used in this file. It should be removed in this commit. It is actually removed in a later commit but ideally it should have been done here.

petebacondarwin commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -697,12 +681,11 @@ export function createTView(
         queries: null,
         viewQuery: viewQuery,
-        node: null!,
+        node: declTNode,
```
🎉

petebacondarwin commented on packages/core/src/render3/interfaces/injector.ts:
```diff
@@ -14,6 +14,11 @@ import {assertDefined, assertEqual} from '../../util/assert';
+/**
+ * Offset from 'injectorIndex' where:
+ * - `TViewData[injectorIndex + TNODE]` => `TNode` associated with the current injector.
+ * - `LView[injectorIndex + TNODE]` => index to the parent injector.
```
Should this comment not be:

```suggestion
 * - `LView[injectorIndex + PARENT_INJECTOR]` => index to the parent injector.
```

And if they need to be the same constant perhaps below it should be:

```ts
export const TNODE = 8;
export const PARENT_INJECTOR = TNODE;
```

petebacondarwin commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -277,11 +277,18 @@ export function createContainerRef(
+        insertView(tView, lView, lContainer, adjustedIdx);
 
+        // Physical operation of adding the DOM nodes.
         const beforeNode = getBeforeNodeForView(adjustedIdx, this._lContainer);
```
Why does this use `this._lContainer` whereas code around it uses `lContainer`, which appears to be the same.

petebacondarwin commented on packages/core/src/render3/view_engine_compatibility.ts:
```diff
@@ -277,11 +277,18 @@ export function createContainerRef(
+        }
 
         (viewRef as ViewRef<any>).attachToViewContainerRef(this);
         addToArray(this._lContainer[VIEW_REFS]!, adjustedIdx, viewRef);
```
Why does this use this._lContainer whereas code around it uses lContainer, which appears to be the same.

petebacondarwin commented on packages/core/src/render3/view_ref.ts:
```diff
@@ -35,11 +35,8 @@ export class ViewRef<T> implements viewEngine_EmbeddedViewRef<T>, viewEngine_Int
-    }
-    return [];
+    const tView = lView[TVIEW];
+    return collectNativeNodes(tView, lView, tView.firstChild, []);
```
🎉

petebacondarwin commented:
Loving the 4th commit :-).

petebacondarwin commented on packages/core/src/render3/component.ts:
```diff
@@ -229,7 +229,7 @@ export function createRootComponent<T>(
   }
 
-  const rootTNode = getPreviousOrParentTNode();
+  const rootTNode = getPreviousOrParentTNode()!;
```
I would argue that wherever you are making this non-null cast, then there should be a dev-mode `assertDefined(getPreviousOrParentTNode())` too.

petebacondarwin commented on packages/core/test/bundling/README.md:
```diff
@@ -6,8 +6,10 @@ This is asserted by keeping gold files of all symbols which are expected to be r
+yarn bazel run --config=ivy //packages/core/test/bundling/forms:symbol_test.accept
 yarn bazel run --config=ivy //packages/core/test/bundling/hello_world:symbol_test.accept
-yarn bazel run --config=ivy //packages/core/test/bundling/injection:symbol_test.accept
+yarn bazel run --config=ivy //packages/core/test/bundling/router:symbol_test.accept
```
Not sure why these changes are in this commit?

petebacondarwin reviewed (approved):
I left a bunch of NITs but nothing blocking. Nice tidy up!

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

IgorMinar commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -41,6 +41,7 @@ export const RENDERER = 11;
 export const SANITIZER = 12;
 export const CHILD_HEAD = 13;
 export const CHILD_TAIL = 14;
+// FIXME(misko): Investigate if the three declarations are really one and the same.
```
```suggestion
// FIXME(misko): Investigate if the three declarations aren't all same thing.
```

mhevery committed:
docs(core): correct instructions for running component repo locally

IgorMinar reviewed (approved):
LGTM 

Reviewed-for: size-tracking

AndrewKushnir commented:
[Presubmit #2](http://test/OCL:331705695:BASE:332343926:1600386340645:cfa07f75).

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -204,25 +205,43 @@ export function getInjectorIndex(tNode: TNode, lView: LView): number {
-    declTNode = view ? view[T_HOST] : null;
-    viewOffset++;
+  let parentTNode: TNode|null = null;
+  let declViewSkips = 0;
```
I would have really benefited from a comment explaining what this variable signifies.

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -204,25 +205,43 @@ export function getInjectorIndex(tNode: TNode, lView: LView): number {
-    view = view[DECLARATION_VIEW]!;
-    declTNode = view ? view[T_HOST] : null;
-    viewOffset++;
+  let parentTNode: TNode|null = null;
```
FWIW I don't believe that the comment above this line is relevant/accurate any more, since skipping this loop would alway result in a `NO_PARENT_INJECTOR`.

petebacondarwin commented on packages/core/src/render3/di.ts:
```diff
@@ -204,25 +205,43 @@ export function getInjectorIndex(tNode: TNode, lView: LView): number {
+    declViewSkips++;
+    if (parentTNode === null) {
+      return NO_PARENT_INJECTOR;
+    } else if (parentTNode.injectorIndex !== -1) {
```
This `else` is not needed here.

And for me it actually complicates understanding, since my brain feels it needs to think about the previous `if` condition when thinking about this `if`.

petebacondarwin replied:
Also when scanning this loop (in fact the whole function) it is hard to see that this is the main path. I.e. that we are primarily aiming to get to the inside of this `if` statement in most scenarios.

petebacondarwin commented on packages/core/src/render3/pipe.ts:
```diff
@@ -48,13 +48,17 @@ export function ɵɵpipe(index: number, pipeName: string): any {
+    setInjectImplementation(previousInjectImplementation);
+  }
   store(tView, getLView(), index, pipeInstance);
   return pipeInstance;
```
If you include the `store()` and `return` statements inside the `try` block then you don't need to use a `let` for pipeInstance.

petebacondarwin commented on packages/core/src/render3/util/injector_utils.ts:
```diff
@@ -5,14 +5,26 @@
+      assertGreaterThan(
+          parentInjectorIndex, HEADER_OFFSET,
+          'Parent injector must be pointing past HEADER_OFFSET.');
   return (parentLocation as any as number) & RelativeInjectorLocationFlags.InjectorIndexMask;
```
```suggestion
  return parentInjectorIndex;
```

I think

AndrewKushnir commented:
[Presubmit #3](http://test/OCL:331705695:BASE:332906901:1600716369980:bfdd5ce0).

jelbourn commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -152,6 +156,10 @@ export const TViewConstructor = class TView implements ITView {
   }
+
+  get type_(): string {
+    return TViewTypeAsString[this.type] || 'TViewType.?' + this.type + '?';
```
```suggestion
    return TViewTypeAsString[this.type] || `TViewType.?${this.type}?`;
```

jelbourn commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -189,23 +197,26 @@ class TNode implements ITNode {
-        return 'TNodeType.View';
-      default:
-        return 'TNodeType.???';
+  debugNodeInjectorPath(lView: LView): DebugNode[] {
```
Add JsDoc description?

(similar for other newly introduced functions)

jelbourn commented on packages/core/src/render3/pipe.ts:
```diff
@@ -48,13 +48,17 @@ export function ɵɵpipe(index: number, pipeName: string): any {
-  setInjectImplementation(previousInjectImplementation);
+  let pipeInstance: any;
+
+  try {
```
Could you add a comment that explains the rationale for using `try`/`catch` here?

mhevery commented:
[presubmit](http://test/OCL:333625996:BASE:333626029:1600991030099:711bcea4)

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -125,32 +102,45 @@ export function createTextNode(value: string, renderer: Renderer3): RText {
-    applyView(tView, lView, renderer, action, renderParent, beforeNode);
-  }
+export function addViewToContainer(
+    tView: TView, renderTNode: TNode, renderer: Renderer3, lView: LView, renderParent: RElement,
```
I'd propose renaming `renderTNode` -> `parentTNode` and may be `renderParent` -> `parentNativeNode` (so both argument names are in sync):
```suggestion
    tView: TView, lView: LView, renderer: Renderer3, parentTNode: TNode, parentNativeNode: RElement,
```

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -382,16 +372,7 @@ export function destroyLView(tView: TView, lView: LView) {
-    // otherwise, use parent view for containers or component views
-    return lViewOrLContainer[PARENT] === rootView ? null : lViewOrLContainer[PARENT];
-  }
+  return lViewOrLContainer[PARENT] === rootView ? null : lViewOrLContainer[PARENT];
```
This function is a one liner now and it's used just once, so we can probably inline this code to the call site (preserving some of the comments that might need an update as well)?

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -529,22 +508,9 @@ function getRenderParent(tView: TView, tNode: TNode, currentView: LView): REleme
   // If the parent tNode is null, then we are inserting across views: either into an embedded view
   // or a component view.
   if (parentTNode == null) {
```
```suggestion
  if (parentTNode === null) {
```

AndrewKushnir commented on packages/core/src/render3/util/discovery_utils.ts:
```diff
@@ -101,7 +102,7 @@ export function getOwningComponent<T>(elementOrDir: Element|{}): T|null {
   ngDevMode && assertLView(lView);
-  while (lView[HOST] === null && (parent = getLViewParent(lView)!)) {
+  while (lView[TVIEW].type === TViewType.Embedded && (parent = getLViewParent(lView)!)) {
     // As long as lView[HOST] is null we know we are part of sub-template such as `*ngIf`
```
This comment needs an update as well to reflect the code or the comment can just be removed as the code is self-explanatory now.

AndrewKushnir commented on packages/core/src/render3/instructions/shared.ts:
```diff
@@ -213,81 +211,58 @@ export function createLView<T>(
-  if (previousOrParentTNode) {
-    if (isParent && previousOrParentTNode.child == null &&
-        (tNode.parent !== null || previousOrParentTNode.type === TNodeType.View)) {
+  if (currentTNode) {
+    if (isParent && currentTNode.child == null && tNode.parent !== null) {
```
```suggestion
  if (currentTNode !== null) {
    if (isParent && currentTNode.child === null && tNode.parent !== null) {
```

AndrewKushnir reviewed (approved):
LGTM 👍 

Just left some comments, but none of them are critical.

mhevery committed:
refactor(core): Create `NodeInjectorOffset` type which better describes NodeInjector

`NodeInjector` is store in expando as a list of values in an array. The
offset constant into the array have been brought together into a single
`NodeInjectorOffset` enum with better documentation explaining their usage.

mhevery committed:
docs(core): Update instructions on updating symbol tests

jelbourn commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -1021,4 +1030,48 @@ export interface DebugNode {
+  /**
+   * A list of Component/Directive instances which were instantiated an this location.
+   */
+  instances: any[];
```
Should this be `unknown[]` instead of `any[]`?

(similar for other newly introduced `any`)

mhevery replied:
fixed

jelbourn commented on packages/core/src/render3/interfaces/view.ts:
```diff
@@ -1021,4 +1030,48 @@ export interface DebugNode {
+  /**
+   * A list of Component/Directive types which need to be instantiated an this location.
+   */
+  factories: Type<any>[];
```
Should this be `Type<unknown>[]`?

mhevery replied:
fixed

jelbourn commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -527,17 +562,72 @@ export function toDebugNodes(tNode: ITNode|null, lView: LView): DebugNode[] {
+ */
+function binary(array: any[], idx: number): string {
+  const value = array[idx];
+  if (typeof value !== 'number') return '????????';
```
Would it make sense to explain here why you return `'????????'` instead of something like `null`?

mhevery replied:
added

jelbourn commented on packages/core/src/render3/instructions/lview_debug.ts:
```diff
@@ -527,17 +562,72 @@ export function toDebugNodes(tNode: ITNode|null, lView: LView): DebugNode[] {
+function binary(array: any[], idx: number): string {
+  const value = array[idx];
+  if (typeof value !== 'number') return '????????';
+  const text = '00000000' + value.toString(2);
```
Add a comment what explains why this isn't just a naive `value.toString(2)`?

mhevery replied:
added

jelbourn reviewed (commented):
Just a few last things I noticed on a second pass

gkalpak reviewed (approved):
Reviewed-for: dev-infra

mhevery commented:
[presubmit](http://test/OCL:333625996:BASE:334195227:1601316184123:30ed099)

alxhub commented:
I'm going to merge this to `master` for the time being. @mhevery, can you please create a separate PR to merge this to the `patch` branch?
